### PR TITLE
Add Density API to ThemeData, implement for buttons.

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This implements a draft API for specifying the concept of `VisualDensity` for a Material user interface.  This will allow us to specify default densities for desktop platforms that are more reasonable, since desktop platforms don't have the same constraints around tap target sizes.

This PR adds the new `VisualDensity` class, and an attribute on the `ThemeData` and `ButtonThemeData` classes for setting it. I applied this new theme information to the buttons which use `RawMaterialButton`. The implementation of the density changes is in `RawMaterialButton`.

## Tests

- Added tests for each of the buttons, and for ThemeData.

## Breaking Change

- [X] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter/lib/src/material/theme_data.dart [hunk 1]
```diff
@@ -1634,3 +1671,109 @@ class _FifoCache<K, V> {
...
+  }
+
+  @override
+  int get hashCode => hashValues(horizontal, vertical, interval);
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DoubleProperty('horizontal', horizontal));
+    properties.add(DoubleProperty('vertical', vertical));
```
Comment (clocksmith) [inline packages/flutter/lib/src/material/theme_data.dart]: interval?

File: packages/flutter/lib/src/material/button.dart [hunk 2]
```diff
@@ -343,9 +355,22 @@ class _RawMaterialButtonState extends State<RawMaterialButton> {
   Widget build(BuildContext context) {
     final Color effectiveTextColor = MaterialStateProperty.resolveAs<Color>(widget.textStyle?.color, _states);
     final ShapeBorder effectiveShape =  MaterialStateProperty.resolveAs<ShapeBorder>(widget.shape, _states);
+    final Offset densityAdjustment = widget.density.adjustment;
+    final BoxConstraints densityConstraints = widget.constraints.copyWith(
+      minWidth: widget.constraints.minWidth != null ? (widget.constraints.minWidth + densityAdjustment.dx).clamp(0.0, double.infinity) : null,
+      minHeight: widget.constraints.minWidth != null ? (widget.constraints.minHeight + densityAdjustment.dy).clamp(0.0, double.infinity) : null,
+    );
+    final EdgeInsetsGeometry padding = widget.padding.add(
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/button.dart]: Could be

```dart
    final EdgeInsetsGeometry padding = _clampInsets(
      widget.padding + EdgeInsets.fromLTRB(
        densityAdjustment.dx,
        densityAdjustment.dy,
        densityAdjustment.dx,
        densityAdjustment.dy,
      ),
    );
```
Or (sorry this is a gratuitous suggestion on top of a gratuitous suggestion)

```dart
    final double dy = densityAdjustment.dx;
    final double dy = densityAdjustment.dy;
    final EdgeInsetsGeometry padding = _clampInsets(widget.padding + EdgeInsets.fromLTRB(dx, dy, dx, dy));
```

**Review by HansMuller — COMMENTED** [review]
Generally speaking this looks fine. A few overall concerns though:

- Should we give Density a less generic name? VisualDensity maybe? It's possible that "Density" will collide with an existing name in some apps.

- It might be best to leave density out of ButtonTheme, and all the rest of the component themes moving forward. At least for now. ButtonTheme includes a colorScheme parameter, which was probably a mistake. Most of the other component themes do not provide properties that shadow generic theme parameters like textTheme or primaryColor. Similarly, specifying density per component also seems like it might be overdoing it, at least at the outset.

- Prefer a gallery demo in a separate PR to a manual test.

**Comment by gspencergoog** [comment]
> - Should we give Density a less generic name? VisualDensity maybe? It's possible that "Density" will collide with an existing name in some apps.

Sure.  I like `VisualDensity`.  Would `UIDensity` work as well?

> - It might be best to leave density out of ButtonTheme, and all the rest of the component themes moving forward. At least for now. ButtonTheme includes a colorScheme parameter, which was probably a mistake. Most of the other component themes do not provide properties that shadow generic theme parameters like textTheme or primaryColor. Similarly, specifying density per component also seems like it might be overdoing it, at least at the outset.

OK, I'll remove it from `ButtonTheme` and have the button only look at the `ThemeData`, and remove it from the per component args.  It will make it harder to set per component, but that might not be too common.

> - Prefer a gallery demo in a separate PR to a manual test.

Actually, I don't think the manual test belongs in the gallery. I mean, there should be a gallery demo for density too (probably just a setting in the overall settings page), but this manual test is really useful as a separate app.

I wrote it because it allows me to take a single component and put it under manual testing where I can actually see it, trying it out in a variety of conditions so that I can quickly see how it behaves under odd conditions that I don't think we want to duplicate in the gallery app. It's also a lot smaller when I have to debug something and look at the widget tree.

For different components, both Darren and I have independently invented similar apps to do what this app does, and it seemed inefficient to keep re-inventing it.

My hope is that these will proliferate, and we'll end up with an easy way to boot up any component and twist the dials and see what happens.

I'm happy to put it into another PR, of course.

**Comment by gspencergoog** [comment]
@HansMuller and I talked about this a bit more, and arrived at having a "profile" on the `VisualDensity` that can be interpreted by the individual components so that they can adjust their own interpretation based on the profile.

They would adjust based on deltas from a "norm" so that instead of using the absolute spec value for a control, it would be a delta from the norm.  Then you could set the norm, and it would be proportional to that.

And, I implemented it here, with `OutlineButton` serving as an example of how customization could be done.

The problem is... you can't lerp an enum, so it blows our ability to lerp smoothly between two "profiles".  Sadness.

**Comment by gspencergoog** [comment]
I think the best thing now might to just rip out the profile stuff entirely, and just associate numbers with the profile names (set up static doubles for the profile types, so `VisualDensity.comfortable` might be -1.0, and `VisualDensity.compact` might be -2.0), and have the components that differ from the norm just have some kind of curve that defines how they map the base density to their own density adjustments.  That keeps the complexity out of the `VisualDensity` object, is infinitely flexible still, and can still be lerped. The downside is that it makes the adjustments at the component level much more complex, but at least that complexity is hidden.

File: packages/flutter/lib/src/material/theme_data.dart [hunk 3]
```diff
@@ -1634,3 +1671,152 @@ class _FifoCache<K, V> {
...
+        assert(vertical != null),
+        assert(interval != null),
+        assert(vertical <= maximumDensity),
+        assert(vertical >= minimumDensity),
+        assert(horizontal <= maximumDensity),
+        assert(horizontal >= minimumDensity);
+
+
+  /// The minimum allowed density.
+  static const double minimumDensity = -4.0;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/theme_data.dart]: Shouldn't this be 
```dart
static const VisualDensity minimum = VisualDensity(horizontal: -4, vertical: -4);
```

The assertions could compare against minimum.vertical, etc

File: packages/flutter/lib/src/material/theme_data.dart [hunk 4]
```diff
@@ -1634,3 +1671,152 @@ class _FifoCache<K, V> {
...
+  /// The horizontal visual density of UI components.
+  ///
+  /// This property affects only the horizontal spacing between and within
+  /// components, to allow for different UI visual densities.
+  ///
+  /// See also:
+  ///
+  ///  - [ThemeData.density], where this property is used to specify the base
+  ///    horizontal density of Material components.
+  ///  - [ButtonThemeData.density], where this property is used to specify the
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/theme_data.dart]: I think we're leaving density out of the component themes.

Here and elsewhere

File: packages/flutter/lib/src/material/theme_data.dart [hunk 5]
```diff
@@ -1634,3 +1671,152 @@ class _FifoCache<K, V> {
...
+        && interval == typedOther.interval;
+  }
+
+  @override
+  int get hashCode => hashValues(horizontal, vertical, interval);
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DoubleProperty('horizontal', horizontal));
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/theme_data.dart]: Aren't default values needed here?

File: packages/flutter_tools/ide_templates/intellij/.idea/runConfigurations/manual_tests___density.xml.copy.tmpl [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<component name="ProjectRunConfigurationManager">
```
Comment (HansMuller) [inline packages/flutter_tools/ide_templates/intellij/.idea/runConfigurations/manual_tests___density.xml.copy.tmpl]: Does this really belong the PR?

Comment (gspencergoog) [reply packages/flutter_tools/ide_templates/intellij/.idea/runConfigurations/manual_tests___density.xml.copy.tmpl]: Yes.  It adds the run configuration for the manual test density app for IntelliJ.

File: packages/flutter/lib/src/material/button.dart [hunk 7]
```diff
@@ -207,6 +208,16 @@ class RawMaterialButton extends StatefulWidget {
   /// The internal padding for the button's [child].
   final EdgeInsetsGeometry padding;
 
+  /// The visual density value for specifying the compactness of button layout.
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/button.dart]: Works for me.

File: packages/flutter/lib/src/material/button.dart [hunk 8]
```diff
@@ -397,7 +421,7 @@ class _RawMaterialButtonState extends State<RawMaterialButton> {
     Size minSize;
     switch (widget.materialTapTargetSize) {
       case MaterialTapTargetSize.padded:
-        minSize = const Size(48.0, 48.0);
+        minSize = Size(48.0 + densityAdjustment.dx, 48.0 + densityAdjustment.dy);
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/button.dart]: Yeah, probably is. Changed.

File: packages/flutter/lib/src/material/theme_data.dart [hunk 9]
```diff
@@ -669,6 +674,32 @@ class ThemeData extends Diagnosticable {
...
+  /// The density value for specifying the compactness of various UI components.
+  ///
+  /// {@template flutter.material.themedata.density}
+  /// Density, in the context of a UI, is the vertical and horizontal
+  /// "compactness" of the elements in the UI. It is unitless, since it means
+  /// different things to different UI elements. For buttons, it affects the
+  /// spacing around the centered label of the button. For lists, it affects the
+  /// distance between baselines of entries in the list.
+  ///
+  /// The values are typically specified in integral values, but any value in
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: Done.

File: packages/flutter/lib/src/material/theme_data.dart [hunk 10]
```diff
@@ -669,6 +674,32 @@ class ThemeData extends Diagnosticable {
...
+  /// range may be used. The range includes values from -4.0 to 4.0, inclusive,
+  /// where negative values display a denser UI, and positive values display a
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: Done.

File: packages/flutter/lib/src/material/theme_data.dart [hunk 11]
```diff
@@ -669,6 +674,32 @@ class ThemeData extends Diagnosticable {
...
+  /// more expanded UI. If a component doesn't support the value given, it will
+  /// clamp to the the nearest value supported.
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: Done.

File: packages/flutter/lib/src/material/theme_data.dart [hunk 12]
```diff
@@ -1634,3 +1671,152 @@ class _FifoCache<K, V> {
     return _cache[key] = loader();
   }
 }
+
+/// Represents the user interface visual density.
+///
+/// Density, in the context of a UI, is the vertical and horizontal
+/// "compactness" of the elements in the UI. It is unitless, since it means
+/// different things to different UI elements.
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: Done.

File: packages/flutter/lib/src/material/theme_data.dart [hunk 13]
```diff
@@ -1634,3 +1671,152 @@ class _FifoCache<K, V> {
     return _cache[key] = loader();
   }
 }
+
+/// Represents the user interface visual density.
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: Done.

File: packages/flutter/lib/src/material/theme_data.dart (hunk 3 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: Nope: "compile time expression depends on itself".

File: packages/flutter/lib/src/material/theme_data.dart (hunk 4 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: Done.

File: packages/flutter/lib/src/material/theme_data.dart (hunk 5 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: Added.

File: packages/flutter/lib/src/material/theme_data.dart [hunk 14]
```diff
@@ -1634,3 +1671,148 @@ class _FifoCache<K, V> {
...
+  /// A const constructor for [VisualDensity].
+  ///
+  /// All of the arguments must be non-null, and [horizontal] and [vertical]
+  /// must be in the interval between [minimumDensity] and [maximumDensity],
+  /// inclusive.
+  const VisualDensity({
+    this.horizontal = 0.0,
+    this.vertical = 0.0,
+    this.interval = 4.0,
+  })  : assert(horizontal != null),
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/theme_data.dart]: nit: 
```suggestion
  }) : assert(horizontal != null),
```
and all following assert statements

File: packages/flutter/lib/src/material/toggle_buttons.dart [hunk 15]
```diff
@@ -819,6 +819,7 @@ class _ToggleButton extends StatelessWidget {
         hoverColor: currentHoverColor,
         splashColor: currentSplashColor,
         focusNode: focusNode,
+        density: theme.density,
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/toggle_buttons.dart]: Is this just to make sure that toggle buttons respond to changes in the theme's density values? In the future, would we need to to allow density to be passed in as a param to `ToggleButtons`?

File: packages/flutter/lib/src/material/outline_button.dart [hunk 16]
```diff
@@ -433,8 +438,16 @@ class _OutlineButtonState extends State<_OutlineButton> with SingleTickerProvide
     ).evaluate(_elevationAnimation);
   }
 
+  // TODO(gspencergoog): implement the real version of this: this is just an
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/outline_button.dart]: Does this just mean that this density value has not been defined by the spec yet?

File: packages/flutter/lib/src/material/button.dart [hunk 17]
```diff
@@ -354,9 +365,22 @@ class _RawMaterialButtonState extends State<RawMaterialButton> {
   Widget build(BuildContext context) {
     final Color effectiveTextColor = MaterialStateProperty.resolveAs<Color>(widget.textStyle?.color, _states);
     final ShapeBorder effectiveShape =  MaterialStateProperty.resolveAs<ShapeBorder>(widget.shape, _states);
+    final Offset densityAdjustment = widget.density.adjustment;
+    final BoxConstraints densityConstraints = widget.constraints.copyWith(
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/button.dart]: Would the variable name `adjustedConstraints` be more appropriate here? In the event that `widget.constraints.minWidth` is null, this variable would not have anything to do with density.

File: packages/flutter/lib/src/material/toggle_buttons.dart (hunk 15 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/toggle_buttons.dart]: Yes, probably.  Removed this for now, so it won't respond to density.

File: packages/flutter/lib/src/material/theme_data.dart [hunk 18]
```diff
@@ -1634,3 +1671,148 @@ class _FifoCache<K, V> {
...
+  /// See also:
+  ///
+  ///  - [ThemeData.density], where this property is used to specify the base
+  ///    vertical density of Material components.
+  final double vertical;
+
+  /// The number of logical pixels represented by an increase or decrease in
+  /// density by 1.0.
+  ///
+  /// This is not always interpreted directly, or even used, by components,
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: I removed this field entirely, in favor of an internal constant.  If people want this control later, we can add it.

File: packages/flutter/lib/src/material/button.dart [hunk 19]
```diff
@@ -397,7 +421,10 @@ class _RawMaterialButtonState extends State<RawMaterialButton> {
     Size minSize;
     switch (widget.materialTapTargetSize) {
       case MaterialTapTargetSize.padded:
-        minSize = const Size(48.0, 48.0);
+        minSize = Size(
+          kMinInteractiveDimension + densityAdjustment.dx,
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/button.dart]: Do we need to clamp these parameters to >= 0?

File: packages/flutter/lib/src/material/theme_data.dart [hunk 20]
```diff
@@ -669,6 +674,34 @@ class ThemeData extends Diagnosticable {
...
+  /// Density, in the context of a UI, is the vertical and horizontal
+  /// "compactness" of the elements in the UI. It is unitless, since it means
+  /// different things to different UI elements. For buttons, it affects the
+  /// spacing around the centered label of the button. For lists, it affects the
+  /// distance between baselines of entries in the list.
+  ///
+  /// Typically, density values are integral, but any value in range may be
+  /// used. The range includes values from [VisualDensity.minimumDensity] (which
+  /// is -4), to [VisualDensity.maximumDensity] (which is 4), inclusive, where
+  /// negative values indicate a denser UI, and positive values indicate a more
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/theme_data.dart]: denser => more compact ?

File: packages/flutter/lib/src/material/theme_data.dart [hunk 21]
```diff
@@ -1635,3 +1674,141 @@ class _FifoCache<K, V> {
...
+    double horizontal,
+    double vertical,
+  }) {
+    return VisualDensity(
+      horizontal: horizontal ?? this.horizontal,
+      vertical: vertical ?? this.vertical,
+    );
+  }
+
+  /// The horizontal visual density of UI components.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/theme_data.dart]: Should we say that the default is 0 is the valid range is -4 to 4? Likewise for horizontal

File: packages/flutter/lib/src/material/theme_data.dart [hunk 22]
```diff
@@ -1635,3 +1674,141 @@ class _FifoCache<K, V> {
...
+///
+/// Density, in the context of a UI, is the vertical and horizontal
+/// "compactness" of the components in the UI. It is unitless, since it means
+/// different things to different UI components.
+///
+/// For example, for buttons, it affects the spacing around the child of the
+/// button. For lists, it affects the distance between baselines of entries in
+/// the list. For chips, it only affects the vertical size, not the horizontal
+/// size.
+class VisualDensity extends Diagnosticable {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/theme_data.dart]: You might mention that the default for the vertical and horizontal density attributes is zero, which means that the component's layout will be just as it appears in the Material Design spec.

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/theme_data.dart (hunk 22 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: I also linked in the material design guidance on density.

File: packages/flutter/lib/src/material/theme_data.dart (hunk 21 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: Done.

**Pull request merged**